### PR TITLE
Force aligned param length <master> [7841]

### PR DESCRIPTION
**Pull request by MiguelCompany**
This is a forward port of #1047 + #1056

### CODE REVIEW

File: include/fastdds/dds/core/policy/QosPolicies.hpp [hunk 1]
```diff
@@ -809,6 +810,7 @@ class GenericDataQosPolicy : public Parameter_t, public QosPolicy,
             const collection_type& vec)
     {
         assign(vec.begin(), vec.end());
+        length = (size() + 7) & ~3;
     }
```
Comment (IkerLuengo) [inline include/fastdds/dds/core/policy/QosPolicies.hpp]: set `hasChanged` to `true`

File: include/fastdds/dds/core/policy/QosPolicies.hpp [hunk 2]
```diff
@@ -828,6 +830,7 @@ class GenericDataQosPolicy : public Parameter_t, public QosPolicy,
...
```
Comment (IkerLuengo) [inline include/fastdds/dds/core/policy/QosPolicies.hpp]: set `hasChanged` to `true`

File: include/fastrtps/qos/QosPolicies.h [hunk 3]
```diff
@@ -62,7 +62,6 @@ constexpr DestinationOrderQosPolicyKind BY_RECEPTION_TIMESTAMP_DESTINATIONORDER_
 constexpr DestinationOrderQosPolicyKind BY_SOURCE_TIMESTAMP_DESTINATIONORDER_QOS =
         DestinationOrderQosPolicyKind::BY_SOURCE_TIMESTAMP_DESTINATIONORDER_QOS;
 using DestinationOrderQosPolicy = fastdds::dds::DestinationOrderQosPolicy;
-using UserDataQosPolicy = fastdds::dds::UserDataQosPolicy;
```
Comment (IkerLuengo) [inline include/fastrtps/qos/QosPolicies.h]: Why treat UserDataQosPolicy different from the rest of QosPolicy classes?

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1552/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1324/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1550/)

File: include/fastrtps/qos/QosPolicies.h (hunk 3 above)
Comment (MiguelCompany) [reply include/fastrtps/qos/QosPolicies.h]: The linker complained on non-defined `UserDataQosPolicy::addToCDRMessage`. This class is only being used inside other classes (ProxyData and DDS QoS), so it is almost impossible for a user to have been using it. In fact, as part of the DDS PSM changes we are making, this has already been done.

File: include/fastdds/dds/core/policy/QosPolicies.hpp [hunk 4]
```diff
@@ -829,8 +833,7 @@ class GenericDataQosPolicy : public Parameter_t, public QosPolicy,
     RTPS_DllAPI inline void setValue(
             const collection_type& vec)
     {
-        assign(vec.begin(), vec.end());
-        length = (size() + 7) & ~3;
+        data_vec(vec);
```
Comment (IkerLuengo) [inline include/fastdds/dds/core/policy/QosPolicies.hpp]: This constructs a local `data_vec`, does not change `this`

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1561/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1333/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1559/)

**Pull request merged**